src/Compilers/Core/VBCSCompiler/AnalyzerConsistencyChecker.cs

```diff
@@ -0,0 +1,104 @@
+        {
+            if (referenceWhiteList == null)
+            {
+                referenceWhiteList = s_defaultWhiteList;
+            }
+
+            try
+            {
+                CompilerServerLogger.Log("Begin Analyzer Consistency Check");
+                return CheckCore(baseDirectory, analyzerReferences, loader, referenceWhiteList);
+            }
+            catch (Exception e)
+            {
+                CompilerServerLogger.LogException(e, "Analyzer Consistency Check");
+                return false;
+            }
+            finally
+            {
+                CompilerServerLogger.Log("End Analyzer Consistency Check");
+            }
+        }
+
+        private static bool CheckCore(string baseDirectory, IEnumerable<CommandLineAnalyzerReference> analyzerReferences, IAnalyzerAssemblyLoader loader, IEnumerable<string> referenceWhiteList)
+        {
+            var resolvedPaths = new List<string>();
+
+            foreach (var analyzerReference in analyzerReferences)
+            {
+                string resolvedPath = FileUtilities.ResolveRelativePath(analyzerReference.FilePath, basePath: null, baseDirectory: baseDirectory, searchPaths: SpecializedCollections.EmptyEnumerable<string>(), fileExists: File.Exists);
+                if (File.Exists(resolvedPath))
+                {
+                    resolvedPath = FileUtilities.TryNormalizeAbsolutePath(resolvedPath);
+                    if (resolvedPath != null)
+                    {
+                        resolvedPaths.Add(resolvedPath);
+                    }
+                }
+
+                // Don't worry about paths we can't resolve. The compiler will report an error for that later.
+            }
+
+            // First, check that the set of references is complete, modulo items in the whitelist.
+            foreach (var resolvedPath in resolvedPaths)
+            {
+                var missingDependencies = AssemblyUtilities.IdentifyMissingDependencies(resolvedPath, resolvedPaths);
+
+                foreach (var missingDependency in missingDependencies)
+                {
+                    if (!referenceWhiteList.Any(name => missingDependency.Name.StartsWith(name)))
+                    {
+                        CompilerServerLogger.Log($"Analyzer assembly {resolvedPath} depends on '{missingDependency}' but it was not found.");
+                        return false;
+                    }
+                }
+            }
+
+            // Second, load all of the assemblies upfront.
+            foreach (var resolvedPath in resolvedPaths)
+            {
+                loader.AddDependencyLocation(resolvedPath);
+            }
+
+            var loadedAssemblies = new List<Assembly>();
+            foreach (var resolvedPath in resolvedPaths)
+            {
+                loadedAssemblies.Add(loader.LoadFromPath(resolvedPath));
```

Comment:
@mavasani There's really no other option. If we waited until the assembly is needed to load and perform these checks, we're already well into the compilation process.